src/rgw/rgw_gc.cc

```diff
@@ -127,11 +127,97 @@ int RGWGC::list(int *index, string& marker, uint32_t max, bool expired_only, std
   return 0;
 }
 
+class RGWGCIOManager {
+  CephContext *cct;
+  RGWGC *gc;
+  int index;
+
+  struct IO {
+    librados::AioCompletion *c{nullptr};
+    string oid;
+    string tag;
+  };
+
+  list<IO> ios;
+  std::list<string> remove_tags;
+
+public:
+  RGWGCIOManager(CephContext *_cct, RGWGC *_gc, int _index) : cct(_cct),
+                                                              gc(_gc),
+                                                              index(_index) {}
+  ~RGWGCIOManager() {
+    for (auto io : ios) {
+      io.c->release();
+    }
+  }
+
+  int schedule_io(IoCtx *ioctx, const string& oid, ObjectWriteOperation *op, const string& tag) {
+#warning configurable
+#define MAX_CONCURRENT_IO 5
+    while (ios.size() > MAX_CONCURRENT_IO) {
+      if (gc->going_down()) {
+        return 0;
+      }
+      handle_next_completion();
+    }
+
+    AioCompletion *c = librados::Rados::aio_create_completion(NULL, NULL, NULL);
+    int ret = ioctx->aio_operate(oid, c, op);
+    if (ret < 0) {
+      return ret;
+    }
+    ios.push_back(IO{c, oid, tag});
+
+    return 0;
+  }
+
+  void handle_next_completion() {
+    assert(!ios.empty());
+    IO& io = ios.front();
+    io.c->wait_for_safe();
+    int ret = io.c->get_return_value();
+    io.c->release();
+
+    if (ret == -ENOENT) {
+      ret = 0;
+    }
+    if (ret < 0) {
+      ldout(cct, 0) << "WARNING: could not remove oid=" << io.oid << ", ret=" << ret << dendl;
+      goto done;
+    }
+
+    remove_tags.push_back(io.tag);
+#define MAX_REMOVE_CHUNK 16
```

Comment:
not too small, not too high... but seriously, this can be now configured, I have no way to tell what is the optimal number. We're bundling this much operations together, more than that (how much more?) seems to me could affect osd availability. Less than that, latency will probably be the biggest factor.